File: v5.4/source/includes/write/transaction.scala
import org.mongodb.scala._
import java.util.concurrent.TimeUnit
import scala.concurrent.Await
import org.mongodb.scala.model.Updates._
import org.mongodb.scala.model.Filters._
import scala.concurrent.duration.Duration
import org.mongodb.scala.result._

object Transaction {
// begin-transaction-method
  def runTransaction(
      database: MongoDatabase,
      observable: SingleObservable[ClientSession]
  ): SingleObservable[ClientSession] = {
    observable.map(clientSession => {
      val moviesCollection = database.getCollection("movies")
      val usersCollection = database.getCollection("users")

      val transactionOptions = TransactionOptions
        .builder()
        .readConcern(ReadConcern.SNAPSHOT)
        .writeConcern(WriteConcern.MAJORITY)
        .build()

      // Starts the transaction with specified options
      clientSession.startTransaction(transactionOptions)

      // Inserts a document into the "movies" collection  
      val insertObservable = moviesCollection.insertOne(
          clientSession,
          Document("name" -> "The Menu", "runtime" -> 107)
      )
      val insertResult = Await.result(insertObservable.toFuture(), Duration(10, TimeUnit.SECONDS))
      println(s"Insert completed: $insertResult")

      // Updates a document in the "users" collection
      val updateObservable = usersCollection.updateOne(
          clientSession,
          equal("name", "Amy Phillips"), set("name", "Amy Ryan")
      )
      val updateResult = Await.result(updateObservable.toFuture(), Duration(10, TimeUnit.SECONDS))
      println(s"Update completed: $updateResult")

      clientSession
    })
  }
// end-transaction-method

  def main(args: Array[String]): Unit = {

    // start-perform-transaction
    val client = MongoClient("<connection string>")
    val database = client.getDatabase("sample_mflix")
    val session = client.startSession();

    val transactionObservable: SingleObservable[ClientSession] =
      runTransaction(database, session)

    val commitTransactionObservable: SingleObservable[Unit] =
      transactionObservable.flatMap(clientSession => clientSession.commitTransaction())

    Await.result(commitTransactionObservable.toFuture(), Duration(10, TimeUnit.SECONDS))
    // end-perform-transaction

    // Keep the main thread alive long enough for the asynchronous operations to complete
    Thread.sleep(1000)
    client.close()
  }

}

File: content/ruby-driver/v2.22/source/includes/write/transaction.rb
require 'bundler/inline'
gemfile do
  source 'https://rubygems.org'
  gem 'mongo'
end

uri = "<connection string URI>"

Mongo::Client.new(uri) do |client|
  #start-txn
  database = client.use('sample_mflix')
  movies_collection = database[:movies]
  users_collection = database[:users]
  
  def run_transaction(session, movies_collection, users_collection)
    transaction_options = {
      read_concern: { level: "snapshot" },
      write_concern: { w: "majority" }
    }
  
  session.with_transaction(transaction_options) do
    # Inserts document into the "movies" collection
    insert_result = movies_collection.insert_one({ name: 'The Menu', runtime: 107 }, session: session)
    puts "Insert completed: #{insert_result.inspect}"
  
    # Updates document in the "users" collection
    update_result = users_collection.update_one({ name: 'Amy Phillips'}, { "$set" => { name: 'Amy Ryan' }}, session: session)
    puts "Update completed: #{update_result.inspect}"
    end
  end
  
  # Starts a session
  session = client.start_session
  
  begin
    # Runs the transaction
    run_transaction(session, movies_collection, users_collection)
    puts "Transaction committed successfully."
  rescue Mongo::Error::OperationFailure => e
    puts "Transaction failed and was aborted. Error: #{e.message}"
  ensure
    session.end_session
  end
  #end-txn
end